library/src/main/java/io/ona/kujaku/services/TrackingService.java

```diff
@@ -0,0 +1,922 @@
+            this.trackingServiceListener.onNewLocationReceived(location);
+        }
+        this.setServiceStatus(TrackingServiceStatus.RUNNING);
+    }
+
+    /**
+     * Inform listener that location registered is close to the departure location
+     *
+     * @param location
+     */
+    private void informCloseToDepartureLocationListener(Location location) {
+        if (this.trackingServiceListener != null && location != null) {
+            if (this.getNumberOfLocationsRecorded() == 1) {
+                return;
+            }
+
+            Location departure = this.getFirstLocationRecorded();
+
+            if (departure != null && departure.distanceTo(location) <= trackingServiceOptions.getDistanceFromDeparture()) {
+                this.trackingServiceListener.onCloseToDepartureLocation(location);
+            }
+        }
+    }
+
+    /**
+     * Get First Location recorded
+     * @return
+     */
+    private Location getFirstLocationRecorded() {
+        if (this.recordedLocations != null && this.recordedLocations.size() > 0) {
+            return this.recordedLocations.get(0);
+        }
+
+        return null;
+    }
+
+    /**
+     * Get number of locations recorded
+     *
+     * @return
+     */
+    private int getNumberOfLocationsRecorded() {
+        if (this.recordedLocations != null) {
+            return this.recordedLocations.size();
+        }
+
+        return 0;
+    }
+
+
+    /**** Public methods that can be used after bind ***/
+
+    /**
+     * Return all locations recorded
+     *
+     * @return
+     */
+    public List<Location> getRecordedLocations() {
+        return this.recordedLocations;
+    }
+
+    /**
+     * Record pending Location
+     */
+    public void takeLocation () {
+        if (pendingRecordingLocation != null) {
+            Location pendingLocation = pendingRecordingLocation;
+            recordPendingLocation();
+            pendingRecordingLocation = pendingLocation;
+        }
+    }
+
+    /**
+     * Class used for the client Binder.  Because we know this service always
+     * runs in the same process as its clients, we don't need to deal with IPC.
+     */
+    public class LocalBinder extends Binder {
+        public TrackingService getService() {
+            // Return this instance of LocalService so clients can call public methods
+            return TrackingService.this;
+        }
+    }
+
+    /*** static methods ***/
+
+    /**
+     * Start Service and bind it
+     *
+     * @param context
+     * @param cls
+     * @param connection
+     * @param options
+     */
+    public static void startAndBindService(Context context, Class<?> cls, ServiceConnection connection, TrackingServiceOptions options) {
+        Intent mIntent = TrackingService.getIntent(context, cls, options);
+        TrackingService.bindService(context, mIntent, connection);
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            context.startForegroundService(mIntent);
```

Comment:
Does this aim to keep the service alive even when the application is not in the foreground? If so, does the service get killed in android versions lower than 0reo?

Reply:
This is the new way of creating a ForegroundService for android > 0reo. But the service himself register as a foreground service too for Android < 0reo with the line "  startForeground(1, notification);"

Reply:
Cool, I guess you mean this 

> If your app targets API level 26 or higher, the system imposes restrictions on using or creating background services unless the app itself is in the foreground. If an app needs to create a foreground service, the app should call startForegroundService()

Reply:
yes